src/index.js

```diff
@@ -150,12 +150,20 @@ class IOpipeWrapperClass {
   invoke() {
     this.runHook('pre:invoke');
     try {
-      return this.userFunc.call(
+      const result = this.userFunc.call(
         this.originalIdentity,
         this.event,
         this.context,
         this.callback
       );
+      if (result && result.then && result.catch) {
+        return new Promise(() => this.callback(null, () => result))
+          .then(value => value)
+          .catch(err => {
+            throw new Error(err);
+          });
+      }
+      return result;
     } catch (err) {
       this.sendReport(err, () => this.originalCallback(err));
       return err;
```

Comment:
I still don't understand what this is doing....? If we're `.catch()` the error in the block above, what is this returning?

Reply:
The enclosing try/catch block should apply to any errors calling this.userFunc.  The new promise/then/catch block should apply only to incoming results that are promises, and catch only if there's an error in there, though as I say, I've not been able to force that catch block to trigger in any way I've been able to monitor yet.

I could see an argument for eliminating the .catch term of the promise/then/catch chain and relying on the enclosing try/catch, but it seems less potentially disruptive to respond to errors in the promise/then/catch the same way the enclosing context does--with the sendReport and returning the error.  The enclosing catch would still apply in cases where result is not a promise.

Reply:
but why are we only returning the `err`? (instead of passing it to a callback/e-throwing)

Reply:
@kolanos I'm not sure why this code was written to return the error instead of throwing (that outer try/catch with return was already in place), though I suspect it may have had to do with making it testable.  There's a test on line 13 of handler.test.js ("Catch typeError by wrapping require block") that fails when I convert this enclosing return to a throw. There may also have been a thought that returning rather than throwing would be less disruptive to the wrapped function, but then...an error is an error, and throwing seems more like what we and AWS would expect.

Reply:
Agree. For testing, not sure why a `expect(func).toThrow()` wasn't used here.

But iopipe-python re-raises uncaught exceptions, so I think we should make this consistent with that behavior. Although it is technically a breaking (or un-breaking?) change, so semver will need to reflect that.

Reply:
The test around this requires *extensive* refactoring.  I think there's value in doing it, but not holding back the feature because of it.  Throwing at the higher level (and correcting the test) could be a separate story.